### ISSUE TITLE
IDC make models load order dependent.

### ISSUE THREAD

**Issue by casperisfine** [Shopify/identity_cache#300]
Currently IDC cause load order issue.

Imagine the following scenario:

```ruby
# app/models/post.rb
class Post < AR::Base
  FOO = 42
  has_many :comments
  cache_has_many :comments
end

# app/models/comment.rb
class Comment < AR::Base
  BAR = Post::FOO + 2
  belongs_to :post
end
```

If you load `Comment` first, what will happen is that `Post` will be loaded while `Comment` is defined but the `belongs_to :post` relation isn't defined yet.

The problem is that `cache_has_many` and `cache_has_one` will invoke methods like [`ensure_cacheable_association`](https://github.com/Shopify/identity_cache/blob/8fdb3d34f0cc632ce2367afb3f08d6fd44ef9587/lib/identity_cache/configuration_dsl.rb#L298-L312) and [`add_parent_expiry_hook`](https://github.com/Shopify/identity_cache/blob/8fdb3d34f0cc632ce2367afb3f08d6fd44ef9587/lib/identity_cache/configuration_dsl.rb#L273-L283) that need the inverse association to be defined. As a result the load will raise an exception.

Another detrimental effect is that in a lazy loading environment (development), referencing a single model will likely lead to all models being loaded.

I searched for ways to avoid this, but without any luck so far.

@sgrif @rafaelfranca any ideas of API that would allow to lazily invoke those methods?

**Comment by rafaelfranca** [comment]
`add_parent_expiry_hook` adds modules to the inverse class, so I think making it lazy load without introducing load order problems is really hard.

The only way that I can think is to stop doing this automatically and ask users to do it manually.

There is also the possibility of inferring the behavior at runtime like it is done with automatic inference of `inverse_of`. We always setup the modules in all models and when executing the code we first check if it should be running it based in some heuristic.

**Comment by casperisfine** [comment]
Indeed. I suppose the only solution, as you say, would be to have to declare the inverse associations explicitly.

It's quite a big of a change.

**Comment by dylanahsmith** [comment]
What if IdentityCache kept track of models it needs to add parent expiry hooks for when could get added to by `cache_has_many` and `cache_has_one` then it could look for those expiry hooks when `include IdentityCache` is called on a model.

The only change needed by users of the library would be to make sure any embedded associations actually `include IdentityCache`, which previously wasn't necessary.  We could check for this when the embedded association is accessed and raise an exception so that we are likely to catch this bug during testing.

**Comment by casperisfine** [comment]
I don't know IDC enough to be sure it wouldn't work right now. But I can try to implement that.

**Comment by casperisfine** [comment]
> What if IdentityCache kept track of models it needs to add parent expiry hooks for when could get added to by cache_has_many and cache_has_one then it could look for those expiry hooks when include IdentityCache is called on a model.

So I tried to pull this of in many ways, but ultimately it's impossible to know the name of the class the association references without loading the model.

For example:
```ruby
class Post < ActiveRecord::Base
  has_many :comments
  cache_has_many :comments
end
```

The referenced class here could be either `Post::Comment` or `Comment`. This ambiguity have rendered all my attempts moot.

I'm afraid the only way to fix this issue would be to either require to declare a `cache_belongs_to` or pass the fully qualified class name in `cache_has_many / cache_has_one`.

That would be quite a huge change.

**Comment by dylanahsmith** [comment]
It still doesn't seem impossible to me, so I'll take a crack at it.

E.g. in the case of a `:comments` association, when `include IdentityCache` is called on either a `Post::Comment` or `Comment` class then `association_reflection.klass` can be used to get the correct constant for installing the after_commit hook.

**Comment by casperisfine** [comment]
Hum, good idea.

**Comment by dylanahsmith** [comment]
Even if we passed the fully qualified class name to `cache_has_many` or `cache_has_one` we would still need to make sure the parent expiry after_commit callback is set before a record on the associated model is modified, so need to at least include something into associated models if they don't already `include IdentityCache`.

We don't want to add extra expirations for these models that we only need cached in an embedded association, which would mean doing

```ruby
include IdentityCache
disable_primary_cache_index
```

in these associated classes that don't already `include IdentityCache`. I would like to simplify that to `include IdentityCache::WithoutPrimaryIndex` which would also allow us to not even define the primary cache index methods.


I have also noticed a bug with autoloading where a parent expiration would be missed if the model that has the embedded association hasn't been loaded yet.  This is more rare with all the forced loading of associations, but would be more likely once we get rid of those.  If we want to avoid that bug, then the parent expiry hook needs to know about the parents to expire before they are loaded, either with an explicit declaration of which parents need to be expired or through inferring it by checking the associated class of cached_belongs_to or belongs_to associations.  This bug wouldn't affect production where models are eagerly loaded.
